### ISSUE TITLE
How should IO operations handle non-unit strided and sparse arrays

### ISSUE THREAD

**Issue by jeremiah-corrado** [chapel-lang/chapel#22076]
There are a few `fileReader`/`fileWriter` methods in the IO module that take array arguments:
 * writeBinary
 * readBinary
 * readLine
 * readAll
 
Currently, these methods only allow one-dimensional, dense, unit-strided arrays. In the future, do we want to support sparse or non-unit-strided arrays? If so, what is the proper way to interpret the array argument's domain?

For example, should the following program print `1 2 3 4 5` or `1 3 5 7 9`?
```chapel
use IO;

const x : [1..10] int = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
      fw = openWriter("test.bin");
fw.writeBinary(x);
fw.close();

const fr = openReader("test.bin");
var D = {1..10},
    d : sparse subdomain(D) = {1, 3, 5, 7, 9},
    y : [d] int;

fr.readBinary(y);
writeln(y);
```

More generally, for the reading methods, should the indices of a sparse domain be used to "index" into a `fileReader`, or should it simply read the next `domain.size` values and disregard the sparsity of the domain? An analogous question applies for arrays with non-unit stride.

Additionally, for `writeBinary`, should a sparse/non-unit-strided array's values be written sequentially, or should the method attempt to seek on the `file` to write values in positions that correspond to the sparse/strided array indices?

**Comment by lydia-duncan** [comment]
I can see both behaviors being useful, personally.  Maybe the way to handle it is for those overloads to provide an additional argument to swap between the two outputs?

**Comment by jeremiah-corrado** [comment]
I agree, I could imagine use cases for both. Something like this would be nice:
```chapel
openReader("mySparseVector.dat").readBinary(sparseArray, useDomainIndices=false);
// or:
openReader("myDenseVector.dat").readBinary(sparseArray, useDomainIndices=true);
```

**Comment by mppf** [comment]
> I can see both behaviors being useful...

I don't think we should add anything here until a user requests it. And even at that point, I might argue that such patterns could/should be written in a different way (e.g. reading elements in a loop, or reading a dense array and copying to a sparse one).

**Comment by bradcray** [comment]
> For example, should the following program print 1 2 3 4 5 or 1 3 5 7 9?

I think definitely 1 2 3 4 5.

> Additionally, for writeBinary, should a sparse/non-unit-strided array's values be written sequentially, or should the method attempt to seek on the file to write values in positions that correspond to the sparse/strided array indices?

Sequentially.

Basically, my answer to both questions is that I think these primitive binary writers should just write out array elements and require the user to do something more clever if they want to preserve indices/structure.

I'd also be OK with not supporting such cases for the time being, though if I were to rank the order in which we should add support for additional cases, I'd say:
* multidimensional dense rectangular, unit-stride arrays
* strided dense rectangular arrays (1D or multidimensional)
* sparse arrays
* distributed arrays (if ever)
